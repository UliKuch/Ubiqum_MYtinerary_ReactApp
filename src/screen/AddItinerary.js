import React from 'react';
import Footer from '../component/Footer';
import Loader from '../component/Loader'
import Navbar from '../component/Navbar'

// redux
import { connect } from "react-redux";
import { findCity } from "../store/actions/cityActions";

// Material-UI
import {
  Box,
  Typography
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

// styles instead of makeStyles in order to use props (cityImage)
const styles = (cityImage) => ({
  cityImage: {
    backgroundImage: `url(${cityImage})`,
  }
});

const useStyles = makeStyles(theme => ({
  cityImage: {
    backgroundPosition: "center",
    backgroundRepeat: "no-repeat",
    backgroundSize: "cover",
    height: 100,
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
  cityName: {
    backgroundColor: theme.palette.background.paper,
    opacity: "0.6",
    width: "100%",
    marginRight: theme.spacing(3),
    marginLeft: theme.spacing(3),
  }
}));


function CityLogo(props) {
  const city = props.city;

  const classes = useStyles();

  return (
    <div>
      <Box
        style={styles(city.img).cityImage}
        className={classes.cityImage}
      >
        <Typography
          variant="h3"
          component="h1"
          className={classes.cityName}
        >
          {city.name}
        </Typography>
      </Box>
    </div>
  )
}

function AddItinerary(props) {
  const cityNamefromUrl = props.match.params.city;

  // fetch city info
  React.useEffect(() => {  
    props.findCity.call(null, cityNamefromUrl,
      window.localStorage.getItem("userToken"))
  }, [cityNamefromUrl, props.findCity])
  


  return (
    <div>
      <Navbar />
      {
        (props.isFetching || !props.city.name)
        ?
        <Loader /> 
        :
        <CityLogo
          city={props.city}
        />
      }

      <Footer />
    </div>
  )
}

function mapStateToProps(state) {
  return {
    city: state.findCity.city,
    isFetching: state.findCity.isFetching
  }
};

const mapDispatchToProps = dispatch => {
  return {
    findCity: (city, token) => dispatch(findCity(city, token))
  }
};

export default connect(
  mapStateToProps, 
  mapDispatchToProps
)(AddItinerary);