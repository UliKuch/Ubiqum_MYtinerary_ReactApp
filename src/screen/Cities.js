import React from 'react';

class Cities extends React.Component {
  render () {
    return (
      null
    )
  }
}

export default Cities;