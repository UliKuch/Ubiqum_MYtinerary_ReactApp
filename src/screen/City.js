import React from 'react';
import Footer from '../component/Footer';
import Itineraries from '../component/Itineraries.js'
import Loader from '../component/Loader'
import Navbar from '../component/Navbar'

// redux
import { connect } from "react-redux";
import { findCity } from "../store/actions/cityActions";

// Material-UI
import { Typography, Box } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

// styles instead of makeStyles in order to use props (cityImage)
const styles = (cityImage) => ({
  cityImage: {
    backgroundImage: `url(${cityImage})`,
  }
});

const useStyles = makeStyles(theme => ({
  cityImage: {
    backgroundPosition: "center",
    backgroundRepeat: "no-repeat",
    backgroundSize: "cover",
    height: 100,
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
  cityName: {
    backgroundColor: theme.palette.background.paper,
    opacity: "0.6",
    width: "100%",
    marginRight: theme.spacing(3),
    marginLeft: theme.spacing(3),
  }
}));

function CityBox(props) {
  const city = props.city;

  const classes = useStyles();
  return (
    <div>
      <Box
        style={styles(city.img).cityImage}
        className={classes.cityImage}
      >
        <Typography
          variant="h3"
          component="h1"
          className={classes.cityName}
        >
          {city.name}
        </Typography>
      </Box>
      <Typography>
        Available MYtineraries:
      </Typography>
      <Itineraries
          cityName={city.name}
      />  
    </div>

  )
}

class City extends React.Component {
  componentDidMount() {
    const cityInput = this.props.match.params.city;
    this.props.findCity(cityInput, window.localStorage.getItem("userToken"));
  }

  render() {
    const city = this.props.city;

    return (
      <div>
        <Navbar />
        {(this.props.isFetching || !city.name) ? <Loader /> 
        : <CityBox
            city={city}
          />
        }
        <a href=".">Choose another city</a>
        <Footer />
      </div>
    )
  }
}

function mapStateToProps(state) {
  return {
    city: state.findCity.city,
    isFetching: state.findCity.isFetching
  }
};

const mapDispatchToProps = dispatch => {
  return {
    findCity: (city, token) => dispatch(findCity(city, token))
  }
};

export default connect(
  mapStateToProps, 
  mapDispatchToProps
)(City);