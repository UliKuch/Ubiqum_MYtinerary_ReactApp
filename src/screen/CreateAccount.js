import React from 'react';
import { Link } from 'react-router-dom';

import Navbar from '../component/Navbar'
import Footer from '../component/Footer'
import GoogleLogin from '../component/GoogleLogin';

// redux
import { connect } from "react-redux";
import { postUser } from "../store/actions/userActions";

// Material-UI
import {
  Avatar,
  Button,
  Grid,
  Typography,
  TextField,
  MenuItem,
  InputLabel,
  Input,
  IconButton,
  InputAdornment,
  FormControl,
  FormControlLabel,
  Checkbox
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import {
  Visibility, 
  VisibilityOff,
  AccountCircle
} from '@material-ui/icons';

// for countries drop down
const countries = [
  {
    value: "France",
    label: "France"
  },
  {
    value: "Germany",
    label: "Germany"
  },
  {
    value: "Netherlands",
    label: "Netherlands"
  },
  {
    value: "Ireland",
    label: "Ireland"
  },
  {
    value: "Spain",
    label: "Spain"
  },
  {
    value: "United Kingdom",
    label: "United Kingdom"
  },
  {
    value: "United States",
    label: "United States"
  },
  {
    value: "Other",
    label: "Other"
  }
]

const useStyles = makeStyles(theme => ({
  textField: {
    margin: theme.spacing(1),
    marginLeft: theme.spacing(3),
    marginRight: theme.spacing(3),
    maxWidth: 300
  },
  formContainer: {
    display: "flex",
    flexDirection: "column"
  },
  checkbox: {
    margin: theme.spacing(1),
    marginLeft: theme.spacing(3),
    marginRight: theme.spacing(3),
  },
  avatar: {
    width: 300,
    height: 300,
    alignSelf: "center"
  },
  submitButton: {
    margin: theme.spacing(1),
    alignSelf: "center",
    width: "100%",
    maxWidth: 300
  }
}));

function CreateAccountForm(props) {
  const classes = useStyles();

  const [values, setValues] = React.useState({
    country: "",
    showPassword: false,
    checked: false,

    // set initial values to avoid uncontrolled components
    userImage: "",
    firstName: "",
    lastName: "",
    username: "",
    password: "",
    email: ""
  });
  
  const handleChangeTextField = name => event => {
    setValues({ ...values, [name]: event.target.value });
  };

  const handleChangeCheckbox = name => event => {
    setValues({ ...values, [name]: event.target.checked });
  };

  const handleClickShowPassword = () => {
    setValues({ ...values, showPassword: !values.showPassword });
  };

  const handleMouseDownPassword = event => {
    event.preventDefault();
  };

  // TODO: add client side validation (pw length, email etc.)

  return (
    <form
      className={classes.formContainer}
      onSubmit={(event) => props.handleSubmit(event, values)}
    >
    {
      values.userImage
      ?
      <Avatar
        src={values.userImage}
        imgProps={{ onError: (e) => { e.target.src = "/images/cross.png" } }}
        className={classes.avatar}
      />
      :
      <Avatar className={classes.avatar}>
        <AccountCircle style={{ fontSize: 300 }}/>
      </Avatar>
    }
      <TextField
        required
        id="UserImage"
        label="User Image"
        placeholder="Enter URL to User Image"
        className={classes.textField}
        value={values.userImage}
        onChange={handleChangeTextField("userImage")}
      />

      <TextField
        required
        id="FirstName"
        label="First Name"
        placeholder="Enter First Name"
        className={classes.textField}
        value={values.firstName}
        onChange={handleChangeTextField("firstName")}
      />

      <TextField
        required
        id="LastName"
        label="Last Name"
        placeholder="Enter Last Name"
        className={classes.textField}
        value={values.lastName}
        onChange={handleChangeTextField("lastName")}
      />

      <TextField
        required
        id="Username"
        label="Username"
        placeholder="Enter Username"
        className={classes.textField}
        value={values.username}
        onChange={handleChangeTextField("username")}
      />

      <FormControl
        required
        id="Password"
        className={classes.textField}
        placeholder="Enter Password"
      >
        <InputLabel htmlFor="adornment-password">Password</InputLabel>
        <Input
          id="adornment-password"
          type={values.showPassword ? "text" : "password"}
          value={values.password}
          onChange={handleChangeTextField("password")}
          endAdornment={
            <InputAdornment position="end">
              <IconButton 
                aria-label="toggle password visibility"
                onClick={handleClickShowPassword}
                onMouseDown={handleMouseDownPassword}
              >
                {values.showPassword ? <Visibility /> : <VisibilityOff />}
              </IconButton>
            </InputAdornment>
          }
        />
      </FormControl>

      <TextField
        required
        id="Email"
        label="E-Mail"
        placeholder="Enter E-Mail"
        className={classes.textField}
        value={values.email}
        onChange={handleChangeTextField("email")}
      />

      <TextField
        required
        select
        id="Country"
        label="Country"
        placeholder="Choose Country"
        className={classes.textField}
        value={values.country}
        onChange={handleChangeTextField("country")}
      >
        {countries.map(option => (
          <MenuItem
            key={option.value}
            value={option.value}
          >
            {option.label}
          </MenuItem>
        ))}
      </TextField>

      <FormControlLabel
        className={classes.checkbox}
        control={
          <Checkbox
            required
            checked={values.checked}
            onChange={handleChangeCheckbox("checked")}
            value="checked"
          />
        }
        label={<label>I agree to MYtinerary's {<Link to="#">
            Terms and Conditions</Link>}.</label>}
      />

      <Button
        variant="contained"
        color="secondary"
        className={classes.submitButton}
        type="submit"
      >
        Submit
      </Button>

    </form>
  )
}


class CreateAccount extends React.Component {
  componentDidMount() {
    // route to landing page if a user is logged in (i.e. if token exists)
    if (window.localStorage.getItem("userToken")) {
      this.props.history.push("/")
    }
  }

  async handleSubmit(event, user) {
    // prevents page reload
    event.preventDefault();

    console.log(user);
    // add user to db and, if successful, log user in
    await this.props.postUser(user)

    // reload after login (logegd in users will be rerouted to landing page)
    window.location.reload();
  }

  render() {

    return (
      <div>
        <Navbar 
          selectedMenuItem="CreateAccount"
        />
        <Grid container direction="column">
          <Grid item container direction="column">
            <Typography variant="h4">Create Account</Typography>
            <CreateAccountForm
              handleSubmit={(event, user) => this.handleSubmit(event, user)}
            />
          </Grid>
          <Typography>or</Typography>
          <GoogleLogin />
        </Grid>
        <Footer />
      </div>
    )
  }
}

function mapStateToProps(state) {

  return {
    isPosting: state.postUser.isPosting,
  }
};

const mapDispatchToProps = dispatch => {
  return {
    postUser: (user) => dispatch(postUser(user)),
  }
};

export default connect(
  mapStateToProps, 
  mapDispatchToProps
)(CreateAccount);