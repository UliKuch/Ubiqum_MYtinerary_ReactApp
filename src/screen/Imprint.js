import React from 'react';

import Navbar from '../component/Navbar';
import Footer from '../component/Footer';

// Material-UI
import {
  Typography,
} from '@material-ui/core';

function Imprint() {
  return (
      <div>
        <Navbar
          selectedMenuItem="Imprint"
        />
        <Typography>
          This is a purely private web page, designed for the purpose of learning how to code only.
          It hence does not need any proper imprint. <br />
          Any images are license free or licensed under creative commons licenses.
        </Typography>
        <Footer />
      </div>
  )
}


export default Imprint;