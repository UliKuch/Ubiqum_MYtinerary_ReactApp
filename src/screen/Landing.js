import React from 'react';
import { Link } from 'react-router-dom';

import Navbar from '../component/Navbar'
import Footer from '../component/Footer'
import Logo from '../component/Logo'

// Carousel
import Carousel from "react-multi-carousel";
import "react-multi-carousel/lib/styles.css";

// redux
import { connect } from "react-redux";
import { fetchCities } from "../store/actions/cityActions";

// Material-UI
import { Container, Typography, Box, CardMedia, Card, CardContent } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(theme => ({
  img: {
    height: "100%"
  },
  arrowImageContainer: {
    height: 100,
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(3),
  },
  text: {
    marginTop: theme.spacing(2),
    marginBottom: theme.spacing(1),
  },
  cityImage: {
    height: 200,
    width: 200
  }
}));

// for carousel
const responsive = {
  desktop: {
    breakpoint: { max: 3000, min: 1024 },
    items: 3,
    slidesToSlide: 3, // optional, default to 1.
  },
  tablet: {
    breakpoint: { max: 1024, min: 464 },
    items: 2,
    slidesToSlide: 2, // optional, default to 1.
  },
  mobile: {
    breakpoint: { max: 464, min: 0 },
    items: 1,
    slidesToSlide: 1, // optional, default to 1.
  },
};

function LandingCarouselImage(props) {
  const classes = useStyles();

  return(
    <Box
      display="flex"
      justifyContent="center"
      component={Link}
      to={`/cities/${props.city.name}`}

    >
      <Card>
        <CardMedia
          image={props.city.img}
          className={classes.cityImage}
        />
        <CardContent>
            <Typography gutterBottom variant="h5" component="h2">
              {props.city.name}
            </Typography>
        </CardContent>
      </Card>
    </Box>
  )
}


function LandingCarousel(props) {
  const images = props.cities.map(city => {
    return (<LandingCarouselImage 
      city={city}
      key={city.name}
      /> )
  })

  return (
    <Carousel
      arrows
      swipeable
      draggable
      showDots={true}
      renderButtonGroupOutside={true}
      responsive={responsive}
      ssr={true} // means to render carousel on server-side.
      infinite={true}
      autoPlay
      autoPlaySpeed={2000}
      keyBoardControl={true}
      transitionDuration={500}
      containerClass="carousel-container"
      // removeArrowOnDeviceType={["tablet", "mobile"]}
      dotListClass="custom-dot-list-style"
      itemClass="carousel-item-padding-40-px"
    >
      {images}
    </Carousel>
  )
}


function LandingMain(props) {
  const classes = useStyles();

  return (
    <Container>
      <Typography
      align="center"
      className={classes.text}
      >
        Find your perfect trip, designed by insiders who know 
            and love their cities.
      </Typography>
      <Box
          className={classes.arrowImageContainer}
        >
        <Box 
          component={Link}
          to="/cities"
        >
          <img
            className={classes.img}
            src="./images/circled-right-2.png"
            alt="Clickable button displaying an arrow."
          />
        </Box>
      </Box>
      <Typography
      align="center"
      className={classes.text}
      >
        Popular MYtineraries:
      </Typography>
      <LandingCarousel
        cities={props.citiesWithImages}
      />
    </Container>
  )
}


class Landing extends React.Component {
  componentDidMount() {
    this.props.fetchCities();

    //TODO: PROBLEM: logs you in if you access landing page with
    // url including token (e.g. if you load page from the urls saved in your browser)

    // if token in url (google login), save token in local storage
    if (this.props.match.params.token) {
      window.localStorage.setItem("userToken", this.props.match.params.token);
    };
  }

  render() {
    // max. 12 cities with images
    let imageCounter = 0;
    const citiesWithImages = this.props.cities.filter(city => {
      if (city.img) {
        imageCounter++;
      }
      return city.img && (imageCounter < 13)
    })

    return (
    <div>
      <Navbar
        selectedMenuItem="Home"
      />
      <Logo />
      <LandingMain
        citiesWithImages={citiesWithImages}
      />
      <Footer />
    </div>
    )
  }
}

function mapStateToProps(state) {
  return {
    cities: state.city.cities,
    isFetching: state.city.isFetching
  }
};

const mapDispatchToProps = dispatch => {
  return {
    fetchCities: () => dispatch(fetchCities()),
  }
};

export default connect(
  mapStateToProps, 
  mapDispatchToProps
)(Landing);