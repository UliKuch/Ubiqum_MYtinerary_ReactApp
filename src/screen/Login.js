import React from 'react';

import Navbar from '../component/Navbar';
import Footer from '../component/Footer';
import GoogleLogin from '../component/GoogleLogin';

// redux
// import { connect } from "react-redux";

// Material-UI
import {
  TextField,
  FormControl,
  InputLabel,
  Input,
  InputAdornment,
  IconButton,
  Button,
  Grid,
  Typography,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import {
  Visibility, 
  VisibilityOff,
} from '@material-ui/icons';

const useStyles = makeStyles(theme => ({
  textField: {
    margin: theme.spacing(1),
    marginLeft: theme.spacing(3),
    marginRight: theme.spacing(3),
    maxWidth: 300
  },
  formContainer: {
    display: "flex",
    flexDirection: "column"
  },
  submitButton: {
    margin: theme.spacing(1),
    alignSelf: "center",
    width: "100%",
    maxWidth: 300
  }
}));

function LoginForm(props) {
  const classes = useStyles();

  const [values, setValues] = React.useState({
    showPassword: false,

    // set initial values to avoid uncontrolled components
    email: "",
    password: "",
  });

  const handleChangeTextField = name => event => {
    setValues({ ...values, [name]: event.target.value });
  };

  const handleClickShowPassword = () => {
    setValues({ ...values, showPassword: !values.showPassword });
  };

  const handleMouseDownPassword = event => {
    event.preventDefault();
  };

  return (
    <form
      className={classes.formContainer}
      onSubmit={(event) => props.handleSubmit(event, values)}
    >
      <TextField
        required
        id="Email"
        label="Email"
        placeholder="Enter Email Address"
        className={classes.textField}
        value={values.email}
        onChange={handleChangeTextField("email")}
      />

      <FormControl
        required
        id="Password"
        className={classes.textField}
        placeholder="Enter Password"
      >
        <InputLabel htmlFor="adornment-password">Password</InputLabel>
        <Input
          id="adornment-password"
          type={values.showPassword ? "text" : "password"}
          value={values.password}
          onChange={handleChangeTextField("password")}
          endAdornment={
            <InputAdornment position="end">
              <IconButton
                aria-label="toggle password visibility"
                onClick={handleClickShowPassword}
                onMouseDown={handleMouseDownPassword}
              >
                {values.showPassword ? <Visibility /> : <VisibilityOff />}
              </IconButton>
            </InputAdornment>
          }
        />
      </FormControl>

      <Button
        variant="contained"
        color="secondary"
        className={classes.submitButton}
        type="submit"
      >
        Submit
      </Button>

    </form>
  )

}

class Login extends React.Component {
  handleSubmit(event, user) {
    // prevents page reload
    event.preventDefault();

    console.log(user);
  }

  render() {
    return (
      <div>
        <Navbar
          selectedMenuItem="Login"
        />
        <Grid container direction="column">
          <Grid item container direction="column">
            <Typography variant="h4">Login</Typography>
            <LoginForm
              handleSubmit={(event, user) => this.handleSubmit(event, user)}
            />
          </Grid>
          <Typography>
          or
          </Typography>
          <GoogleLogin />
        </Grid>
        <Footer />
      </div>
    )
  }

}

export default Login