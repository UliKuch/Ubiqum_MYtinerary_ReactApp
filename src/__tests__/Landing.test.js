import React from 'react';
import { shallow } from 'enzyme';
import { Landing } from '../screen/Landing';
import toJson from 'enzyme-to-json';

it('should render without crashing', () => {

  const match = { params: { token: "" } };
  const cities = [{}];
  const isFetching = false;

  shallow(<Landing match={match} cities={cities} isFetching={isFetching} />);
});

test('Landing component snapshot', () => {
  const match = { params: { token: "" } };
  const cities = [{}];
  const isFetching = false;

  const wrapper = shallow(<Landing match={match} cities={cities} isFetching={isFetching} />);

  expect(toJson(wrapper)).toMatchSnapshot();
});