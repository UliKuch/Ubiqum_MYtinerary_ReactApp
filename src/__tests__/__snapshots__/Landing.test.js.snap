// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing component snapshot 1`] = `
<div>
  <Connect(Navbar)
    selectedMenuItem="Home"
  />
  <Logo />
  <WithStyles(ForwardRef(Container))>
    <WithStyles(ForwardRef(Typography))
      align="center"
      className="makeStyles-text-3"
    >
      Find your perfect trip, designed by insiders who know and love their cities.
    </WithStyles(ForwardRef(Typography))>
    <Styled(MuiBox)
      className="makeStyles-arrowImageContainer-2"
    >
      <Styled(MuiBox)
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Link",
            "propTypes": Object {
              "innerRef": [Function],
              "onClick": [Function],
              "replace": [Function],
              "target": [Function],
              "to": [Function],
            },
            "render": [Function],
          }
        }
        to="/cities"
      >
        <img
          alt="Clickable button displaying an arrow."
          className="makeStyles-img-1"
          src="/images/circled-right-2.png"
        />
      </Styled(MuiBox)>
    </Styled(MuiBox)>
    <WithStyles(ForwardRef(Typography))
      align="center"
      className="makeStyles-text-3"
    >
      Popular MYtineraries:
    </WithStyles(ForwardRef(Typography))>
    <LandingCarousel
      citiesSubarrays={Array []}
      isFetching={false}
    />
  </WithStyles(ForwardRef(Container))>
  <Footer />
</div>
`;
